### PR TITLE
Patch Api for Encryption

### PR DESCRIPTION
**Pull request by aayush3011**
1. Added Patch Api for Encryption.
2. Added a test case to test the functionality.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java [hunk 1]
```diff
@@ -257,28 +257,44 @@ public EncryptionSettings getEncryptionSettings() {
...
-                        }
-                        return Mono.empty();
-                    });
-                    encryptionMonoList.add(voidMono);
+                if (itemJObj.isValueNode()) {
+                    if (patchPropertyPath.substring(1).equals(propertyName)) {
+                        return this.encryptionSettings.getEncryptionSettingForPropertyAsync(propertyName,
+                            this).flatMap(settings -> {
+                                try{
+                                    return Mono.just(EncryptionUtils.getSimpleObjectMapper().readTree(EncryptionUtils.getSimpleObjectMapper().writeValueAsString(encryptAndSerializeValue(settings,
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java:266]: Check id Jackson give direct api for byte[] to JsonNode

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java [hunk 2]
```diff
@@ -257,28 +257,44 @@ public EncryptionSettings getEncryptionSettings() {
...
-                        this).flatMap(settings -> {
-                        try {
-                            encryptAndSerializeProperty(settings, itemJObj, propertyValueHolder, propertyName);
-                        } catch (MicrosoftDataEncryptionException ex) {
-                            return Mono.error(ex);
-                        }
-                        return Mono.empty();
-                    });
-                    encryptionMonoList.add(voidMono);
+                if (itemJObj.isValueNode()) {
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java:261]: Do patch api allow to update non primitive types, like a composite object inside item , if yes then we need to handle that scenario

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java [hunk 3]
```diff
@@ -530,6 +537,101 @@ EncryptionProcessor getEncryptionProcessor() {
...
+                                                                    boolean isRetry) {
+
+        setRequestHeaders(requestOptions);
+        return this.container.patchItem(itemId, partitionKey, encryptedCosmosPatchOperations, requestOptions, itemType).publishOn(encryptionScheduler).
+            flatMap(cosmosItemResponse -> setByteArrayContent((CosmosItemResponse<byte[]>)cosmosItemResponse,
+                this.encryptionProcessor.decrypt(this.cosmosItemResponseBuilderAccessor.getByteArrayContent((CosmosItemResponse<byte[]>)cosmosItemResponse)))
+                .map(bytes -> this.responseFactory.createItemResponse((CosmosItemResponse<byte[]>) cosmosItemResponse,
+                    itemType))).onErrorResume(exception -> {
+                        if(!isRetry && exception instanceof CosmosException) {
+                            this.encryptionProcessor.getIsEncryptionSettingsInitDone().set(false);
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java:627]: isIncorrectContainerRid(cosmosException) condition missing , please check other api

File: sdk/cosmos/azure-cosmos-encryption/src/test/java/com/azure/cosmos/encryption/EncryptionAsyncApiCrudTest.java [hunk 4]
```diff
@@ -526,6 +526,56 @@ public void invalidDataEncryptionKeyAlgorithm() {
         }
     }
 
+    @Test(groups = {"encryption"}, timeOut = TIMEOUT)
+    public void patchItem() {
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos-encryption/src/test/java/com/azure/cosmos/encryption/EncryptionAsyncApiCrudTest.java:530]: Please add patch state cache test in test "crudQueryStaleCache", check  how we doing other operation including batch

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java [hunk 5]
```diff
@@ -242,11 +242,50 @@ public EncryptionSettings getEncryptionSettings() {
...
+                            }
+                        });
+                    }
+                }
+            }
+            return Mono.empty();
+        }));
+    }
+
+    public Mono<JsonNode> encryptObjectNode(JsonNode itemJObj, String patchPropertyPath) {
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java:288]: Please remove patchPropertyPath, not getting used

**Review by simplynaveen20 — APPROVED** [review]
LGTM , please address comments as discussed over call.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java (hunk 1 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java]: Could not find a simpler way of doing it.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java (hunk 2 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java]: Added the code for composite patch.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java [hunk 6]
```diff
@@ -530,6 +537,101 @@ EncryptionProcessor getEncryptionProcessor() {
...
+     *
+     * @return an {@link Mono} containing the Cosmos item resource response with the patched item or an error.
+     */
+    public <T> Mono<CosmosItemResponse<T>> patchItem(
+        String itemId,
+        PartitionKey partitionKey,
+        CosmosPatchOperations cosmosPatchOperations,
+        CosmosPatchItemRequestOptions options,
+        Class<T> itemType) {
+
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java:561]: Done

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java (hunk 3 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java]: Added the check

File: sdk/cosmos/azure-cosmos-encryption/src/test/java/com/azure/cosmos/encryption/EncryptionAsyncApiCrudTest.java (hunk 4 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos-encryption/src/test/java/com/azure/cosmos/encryption/EncryptionAsyncApiCrudTest.java:563]: Added the test

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java (hunk 5 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/implementation/EncryptionProcessor.java]: Removed

**Pull request merged**